### ISSUE TITLE
"nan" behaves differently in CPU and GPU

### ISSUE THREAD

**Issue by ShigekiKarita** [pytorch/pytorch#1810]
Should I report this to torch repository?

``` python
>>> torch.topk(torch.FloatTensor([-1,float("nan"),2, 1]), 1)
(
  2
 [torch.FloatTensor of size 1], 
  2
 [torch.LongTensor of size 1])

>>> torch.topk(torch.cuda.FloatTensor([-1,float("nan"),2]), 1)
(
 nan
 [torch.cuda.FloatTensor of size 1 (GPU 0)], 
  9.2054e+18
 [torch.cuda.LongTensor of size 1 (GPU 0)])
```

In CPU, `topk` seems to ignore "nan" and its indices from result.
In GPU, `topk` with "nan" is kind of undefined behavior?

and other examples 

``` python
>>> torch.max(torch.cuda.FloatTensor([float("nan"), 1]))
1.0

>>> torch.max(torch.FloatTensor([float("nan"), 1]))
nan

>>> torch.min(torch.cuda.FloatTensor([float("nan"), 1]))
1.0

>>> torch.min(torch.FloatTensor([float("nan"), 1]))
nan
```

``` python
>>> torch.sort(torch.FloatTensor([float("nan"), 1]))
(
 nan
   1
 [torch.FloatTensor of size 2], 
  0
  1
 [torch.LongTensor of size 2])

>> torch.sort(torch.cuda.FloatTensor([float("nan"), 1]))
(
   1
 nan
 [torch.cuda.FloatTensor of size 2 (GPU 0)], 
  1
  0
 [torch.cuda.LongTensor of size 2 (GPU 0)])
```

cc @VitalyFedyunin @ngimel @heitorschueroff

**Comment by ShigekiKarita** [comment]
I found some cutorch issues on nan already reported
+ Inconsistent handling of NaNs in min() and max() on Cuda vs CPU tensors https://github.com/torch/cutorch/issues/289
+ Topk function get wired index https://github.com/torch/cutorch/issues/377
+ torch.multinomial should fail on nan input https://github.com/torch/cutorch/issues/509

Does anyone try torch.sort with nan on lua torch?

**Comment by vadimkantorov** [comment]
Just encountered the topk issue too

**Comment by apaszke** [comment]
So my recommendation would be to never work with NaNs (if you have some then you probably have a bug). I don't think it's worth keeping CPU and GPU NaN handling consistent if there's any possibility that it might decrease performance.

**Comment by vadimkantorov** [comment]
Sure thing, it was just strange to get "infinite" indices out of topk that cause strange error or crashes downstream

**Comment by vadimkantorov** [comment]
workaround may be a standard way to register a hook that checks outputs of all computation graph edges for NaNs and breaks if there are any, then there wouldn't be bad surprises

**Comment by 0wu** [comment]
Wouldn't it be beneficial to propagate NaNs to help discover problems in code? For example DivisionByZero and Overflows. Also, nans provide to warn a deep-net trainer that their learning rate might be too large to cause internal overflows.

We can certainly make a flag to decide if we want to include nans, (like how Matlab did it https://www.mathworks.com/help/matlab/ref/min.html#inputarg_nanflag)

**Comment by apaszke** [comment]
It would be beneficial, but CPUs and GPUs are devices with different performance characteristics, which means that the fastest algorithms for them are likely to be different, and ways they propagate NaNs might be different as well. I agree that we might want to propagate NaNs, but if the tradeoff for this are dramatic performance regressions then I don't think it's worth it.

**Comment by ezyang** [comment]
We looked at this in triage: the important thing is for cpu and gpu to be consistent, and for us to be consistent across operators wrt how NaNs compare. It's not the worst thing in the world for NaNs to show up in topk if they are there.

cc @umanwizard

**Comment by vadimkantorov** [comment]
For me the problem with topk were the crazy large negative indices. Maybe better to reset them to a documented special values or to the actual indices, corresponding to nan positions (if they are selected)

**Comment by ngimel** [comment]
Closing, a lot of these things are fixed, please open a separate issue if some of the behaviors are still unexpected.
